add gradle tasks for metalava generation

VinayGuthal opened the pull request:
This pr does the following things. it provides two tasks
1. generateApiTxt task
This generates the api.txt file which can be used to generate api information in the future.
2. apiInformation task
This provides information about the public apis.
Adds the ci command to invoke the github rest api to post the comments of the results.

Additional context:
Baseline files are certain files which allows metalava to ignore the warnings present in these files. There is a system property updateBaseline which is added which allows us to control updation of the baseline file.

The last comment is how the output would look like

VinayGuthal commented:
Can you guys have an initial look onto this pr

VinayGuthal commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -64,6 +69,21 @@ public void apply(Project project) {
     }
+    if(System.getenv().containsKey("METALAVA_BINARY_PATH")) {
+      String metalavaBinaryPath = System.getenv().get("METALAVA_BINARY_PATH");
+      String sourcePath = project.getProjectDir() + "/src/main/java";
```
probably needs different paths if its a kotlin project

vkryachko replied:
This needs to be replaced with the variant-specific sourceSets collection: see https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-master-dev/buildSrc/src/main/kotlin/androidx/build/java/JavaCompileInputs.kt#43

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.groovy:
```diff
@@ -0,0 +1,48 @@
+
+
+@CompileStatic
+class ApiInformationTask extends DefaultTask {
```
Can you change this to `.java`?

vkryachko replied:
Also pls add javadoc to the class, it was unclear to me what it does

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/GenerateApiTask.groovy:
```diff
@@ -0,0 +1,43 @@
+
+
+@CompileStatic
+class GenerateApiTask extends DefaultTask {
```
`.java` as well

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a testing comment

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
[34m                _        _
 _ __ ___   ___| |_ __ _| | __ ___   ____ _
| '_ ` _ \ / _ \ __/ _` | |/ _` \ \ / / _` |
| | | | | |  __/ || (_| | | (_| |\ V / (_| |
|_| |_| |_|\___|\__\__,_|_|\__,_| \_/ \__,_|[0m

[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentId.java:23: [31merror: [0mClass com.google.firebase.firestore.DocumentId no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/Exclude.java:23: [31merror: [0mClass com.google.firebase.firestore.Exclude no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/IgnoreExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.IgnoreExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/PropertyName.java:23: [31merror: [0mClass com.google.firebase.firestore.PropertyName no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ServerTimestamp.java:23: [31merror: [0mClass com.google.firebase.firestore.ServerTimestamp no longer implements java.lang.annotation.Annotation [RemovedInterface]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/ThrowOnExtraProperties.java:23: [31merror: [0mClass com.google.firebase.firestore.ThrowOnExtraProperties no longer implements java.lang.annotation.Annotation [RemovedInterface]

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
This is a test comment

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
.[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:29: [31merror: [0mAdded class com.google.firebase.firestore.DocumentChange [AddedClass]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:69: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:95: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getDocument() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:117: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getNewIndex() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:108: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getOldIndex() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:90: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getType() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:81: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.hashCode() [AddedMethod]

The public api surface has changed for the subproject firebase-common:
.[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/api.txt:36: [31merror: [0mRemoved method com.google.firebase.FirebaseOptions.helloWorld() [RemovedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/src/main/java/com/google/firebase/FirebaseTooManyRequestsException.java:25: [31merror: [0mAdded constructor com.google.firebase.FirebaseTooManyRequestsException(String) [AddedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
.[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:29: [31merror: [0mAdded class com.google.firebase.firestore.DocumentChange [AddedClass]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:69: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:95: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getDocument() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:117: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getNewIndex() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:108: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getOldIndex() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:90: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getType() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:81: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.hashCode() [AddedMethod]

The public api surface has changed for the subproject firebase-common:
.[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/api.txt:36: [31merror: [0mRemoved method com.google.firebase.FirebaseOptions.helloWorld() [RemovedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/src/main/java/com/google/firebase/FirebaseTooManyRequestsException.java:25: [31merror: [0mAdded constructor com.google.firebase.FirebaseTooManyRequestsException(String) [AddedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
.[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:29: [31merror: [0mAdded class com.google.firebase.firestore.DocumentChange [AddedClass]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:69: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:95: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getDocument() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:117: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getNewIndex() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:108: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getOldIndex() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:90: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.getType() [AddedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:81: [31merror: [0mAdded method com.google.firebase.firestore.DocumentChange.hashCode() [AddedMethod]

The public api surface has changed for the subproject firebase-common:
.[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/api.txt:36: [31merror: [0mRemoved method com.google.firebase.FirebaseOptions.helloWorld() [RemovedMethod]
[1m/usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/src/main/java/com/google/firebase/FirebaseTooManyRequestsException.java:25: [31merror: [0mAdded constructor com.google.firebase.FirebaseTooManyRequestsException(String) [AddedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
./usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:29: error: Added class com.google.firebase.firestore.DocumentChange [AddedClass]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:69: error: Added method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:95: error: Added method com.google.firebase.firestore.DocumentChange.getDocument() [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:117: error: Added method com.google.firebase.firestore.DocumentChange.getNewIndex() [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:108: error: Added method com.google.firebase.firestore.DocumentChange.getOldIndex() [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:90: error: Added method com.google.firebase.firestore.DocumentChange.getType() [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:81: error: Added method com.google.firebase.firestore.DocumentChange.hashCode() [AddedMethod]

The public api surface has changed for the subproject firebase-common:
./usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/api.txt:36: error: Removed method com.google.firebase.FirebaseOptions.helloWorld() [RemovedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/src/main/java/com/google/firebase/FirebaseTooManyRequestsException.java:25: error: Added constructor com.google.firebase.FirebaseTooManyRequestsException(String) [AddedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
./usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:29: error: Added class com.google.firebase.firestore.DocumentChange [AddedClass]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:69: error: Added method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:95: error: Added method com.google.firebase.firestore.DocumentChange.getDocument() [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:117: error: Added method com.google.firebase.firestore.DocumentChange.getNewIndex() [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:108: error: Added method com.google.firebase.firestore.DocumentChange.getOldIndex() [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:90: error: Added method com.google.firebase.firestore.DocumentChange.getType() [AddedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentChange.java:81: error: Added method com.google.firebase.firestore.DocumentChange.hashCode() [AddedMethod]

The public api surface has changed for the subproject firebase-common:
./usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/api.txt:36: error: Removed method com.google.firebase.FirebaseOptions.helloWorld() [RemovedMethod]
/usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/src/main/java/com/google/firebase/FirebaseTooManyRequestsException.java:25: error: Added constructor com.google.firebase.FirebaseTooManyRequestsException(String) [AddedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
No project's public api surface was changed due to this PR.

VinayGuthal commented:
/retest

VinayGuthal commented:
No project's public api surface was changed due to this PR.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-common:
.error: Added class com.google.firebase.FirebaseOptions.Builder [AddedClass]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-common:
error: Added class com.google.firebase.FirebaseOptions.Builder [AddedClass]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
No project's public api surface was changed due to this PR.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
error: Added class com.google.firebase.firestore.DocumentChange [AddedClass]
error: Added method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getDocument() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getNewIndex() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getOldIndex() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getType() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.hashCode() [AddedMethod]

The public api surface has changed for the subproject firebase-common:
error: Removed method com.google.firebase.FirebaseOptions.helloWorld() [RemovedMethod]
error: Added constructor com.google.firebase.FirebaseTooManyRequestsException(String) [AddedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
error: Added class com.google.firebase.firestore.DocumentChange [AddedClass]
error: Added method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getDocument() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getNewIndex() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getOldIndex() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getType() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.hashCode() [AddedMethod]

The public api surface has changed for the subproject firebase-common:
error: Removed method com.google.firebase.FirebaseOptions.helloWorld() [RemovedMethod]
error: Added constructor com.google.firebase.FirebaseTooManyRequestsException(String) [AddedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
error: Added class com.google.firebase.firestore.DocumentChange [AddedClass]
error: Added method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getDocument() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getNewIndex() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getOldIndex() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.getType() [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.hashCode() [AddedMethod]

The public api surface has changed for the subproject firebase-common:
error: Removed method com.google.firebase.FirebaseOptions.helloWorld() [RemovedMethod]
error: Added constructor com.google.firebase.FirebaseTooManyRequestsException(String) [AddedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
No project's public api surface was changed due to this PR.

VinayGuthal commented:
No project's public api surface was changed due to this PR.

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +61,40 @@ public void apply(Project project) {
+        sourcePaths.add(directory.getAbsolutePath());
+      }
+      String sourcePathArgument = String.join(",", sourcePaths);
+      project.getTasks().create("apiInformation", ApiInformationTask.class, task -> {
```
You should prefer `.register()` instead of `.create()` as it is evaluated lazily.

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java:
```diff
@@ -0,0 +1,98 @@
+
+
+    private String getCmdOutput (String cmdToRun) {
+        ProcessBuilder processBuilder = new ProcessBuilder(cmdToRun.split(" "));
```
you should be able to just use `getProject().javaexec(spec -> {/*configure javaexec*/});`

vkryachko replied:
or alternatively just plain `getProject().exec()`

vkryachko replied:
this is still not done

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java:
```diff
@@ -0,0 +1,98 @@
+
+    }
+
+    private void writeToFile(String output) {
```
with exec/javaexec you should be able to write stdout/stderr directly into the file if you like.

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/GenerateApiTask.java:
```diff
@@ -0,0 +1,52 @@
+/**
+    Task generates the api txt file for the current source code.
+ */
+@CompileStatic
```
Same comments apply as for the task above

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java:
```diff
@@ -0,0 +1,98 @@
+        String cmdToRun = String.format(cmdTemplate, metalavaBinaryPath, sourcePath, apiTxt, baselinePath);
+        String cmdOutput = getCmdOutput(cmdToRun);
+        if(cmdOutput == null ){
+            System.out.println("Unable to run the command " + cmdToRun);
```
You should not write to stdout directly instead use `getLogger().error()` or any other log level.(Applies throughout the PR)

vkryachko commented on ci/fireci/fireci/commands.py:
```diff
@@ -61,3 +63,38 @@ def smoke_tests(app_build_variant, test_apps_dir):
+
+@click.argument('issue_number', required=True)
+@click.argument('repo_name', required=True)
+@click.argument('auth_token', required=True)
```
Consider making all of the above `option()`s instead of arguments, to make the cli invocation more readable.
Additionally you can set `envvar` on each of the following options to make them optional:
issue_number: `envvar="PULL_NUMBER"`
repo_name: `envvar="REPO_NAME"`

vkryachko commented on ci/fireci/fireci/commands.py:
```diff
@@ -61,3 +63,38 @@ def smoke_tests(app_build_variant, test_apps_dir):
+  """Comments the api information on the pr"""
+
+  gradle.run('apiInformation')
+  gradle.run('generateApiTxtFile')
```
you can invoke them in one gradle invocation, e.g. `gradle.run("task1", "task2")`.

Also you have 3 java tasks defined above, but here you're invoking only 2, is one of them not needed?

VinayGuthal replied:
I have only 2 tasks defined above?

vkryachko replied:
I can see 3 :)

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +61,40 @@ public void apply(Project project) {
+        task.setProperty("apiTxt", project.getProjectDir() + "/api.txt");
+        task.setProperty("metalavaBinaryPath", METALAVA_BINARY_PATH);
+        task.setProperty("sourcePath", sourcePathArgument);
+        task.setProperty("outputPath", project.getRootProject().getBuildDir() + "/apiinfo/subproject" + project.getPath());
```
can you name files something like "project_path_name_api.txt" instead of "subproject:project:path:name"? It's not clear to me why you need the "subproject" prefix, and colons in paths need a lot of escaping

vkryachko commented on ci/fireci/fireci/commands.py:
```diff
@@ -61,3 +63,38 @@ def smoke_tests(app_build_variant, test_apps_dir):
+    comment_string += ('Please update the api.txt files for the subprojects being affected by this change '
+      'with the files present in the artifacts directory. Also perform a major/minor bump accordingly.\n')
+  else:
+    comment_string = 'No project\'s public api surface was changed due to this PR.\n'
```
Do we want to post the comment unconditionally? It might make sense to not post it at all if everything looks ok, wdyt?

VinayGuthal commented:
No project's public api surface was changed due to this PR.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-database:
error: Added class com.google.firebase.database.DataSnapshot [AddedClass]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
error: Removed method com.google.firebase.firestore.DocumentChange.val(Object) [RemovedMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
/retest

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +62,45 @@ public void apply(Project project) {
+    if (System.getenv().containsKey("METALAVA_BINARY_PATH")) {
+      String metalavaBinaryPath = System.getenv("METALAVA_BINARY_PATH");
+      AndroidSourceSet mainSourceSet = android.getSourceSets().getByName("main");
+      File apiTxtFile = new File(project.getProjectDir() + "/api.txt");
```
You can replace this with project.file("api.txt"), and maybe inline it?

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +62,45 @@ public void apply(Project project) {
+      String metalavaBinaryPath = System.getenv("METALAVA_BINARY_PATH");
+      AndroidSourceSet mainSourceSet = android.getSourceSets().getByName("main");
+      File apiTxtFile = new File(project.getProjectDir() + "/api.txt");
+      File baseLineFile = new File(project.getProjectDir() + "/baseline.txt");
```
What's the purpose of baseline.txt? it seems to be empty, also why is it's version declared as 1.0, while api.txt is declared as 2.0?

VinayGuthal replied:
Baseline files are certain files which allows metalava to ignore the warnings present in these files. There is a system property updateBaseline which is added which allows us to control updation of the baseline file.

VinayGuthal replied:
These files are empty for some sdks whereas its not for firestore, common and other sdks

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +62,45 @@ public void apply(Project project) {
+      AndroidSourceSet mainSourceSet = android.getSourceSets().getByName("main");
+      File apiTxtFile = new File(project.getProjectDir() + "/api.txt");
+      File baseLineFile = new File(project.getProjectDir() + "/baseline.txt");
+      File outputFile = new File(project.getRootProject().getBuildDir() + "/apiinfo/" + project.getPath().substring(1).replace(":", "_"));
```
```java
project.getRootProject().file(Paths.get(
  project.getRootProject().buildDir().getPath(),
  "apiinfo",
  project.getPath().substring(1).replace(":", "_")));
```

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +62,45 @@ public void apply(Project project) {
+      File baseLineFile = new File(project.getProjectDir() + "/baseline.txt");
+      File outputFile = new File(project.getRootProject().getBuildDir() + "/apiinfo/" + project.getPath().substring(1).replace(":", "_"));
+      File outputFileDir = outputFile.getParentFile();
+      if(!outputFileDir.exists()) {
```
You should move it into the task action so that this does not happen if the task is not executed.

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java:
```diff
@@ -0,0 +1,100 @@
+ */
+public class ApiInformationTask extends DefaultTask {
+
+    @Input
```
see: https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/Input.html

specifically:
```
This annotation should be attached to the getter method in Java or the property in Groovy.
Annotations on setters or just the field in Java are ignored.
```

vkryachko replied:
Since that is the case, I;d suggest you make the changes I recommended above(re `abstract` task)

VinayGuthal commented:
/retest

VinayGuthal commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +62,45 @@ public void apply(Project project) {
+        task.setProperty("sourcePath", sourcePathArgument);
+        task.setProperty("outputFile", outputFile);
+        task.setProperty("baselineFile", baseLineFile);
+        if (project.hasProperty("updateBaseline")) {
```
nope. Bcz sometimes we wouldnt want to update the baseline file even if it exists

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
error: Added method com.google.firebase.firestore.CollectionReference.add(Object) [AddedMethod]
error: Added method com.google.firebase.firestore.DocumentChange.equals(Object) [AddedMethod]
Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-database:
error: Added class com.google.firebase.database.ChildEventListener [AddedInterface]
error: Added class com.google.firebase.database.DatabaseException [AddedClass]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

ashwinraghav commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java:
```diff
@@ -0,0 +1,93 @@
+        }
+        String cmdTemplate = "%s --source-path %s --check-compatibility:api:current %s --format=v2 --baseline %s --no-color";
+        if(getUpdateBaseline()) {
+            cmdTemplate = "%s --source-path %s --check-compatibility:api:current %s --format=v2 --update-baseline %s --no-color";
```
IMO, there is a lot of boilerplate here that comes down to 2 commands and their respective arguments. 
```
%s --source-path %s --api %s --format=v2 --baseline %s
```
```
 "%s --source-path %s --check-compatibility:api:current %s --format=v2 --update-baseline %s --no-color";
```
I think we can remove some cruft by modelling this as 
```java
// This task executes in 2 modes.
// 1. Generate api text mode
// 2. Compare api text mode 
public abstract class MetalavaExecutionTask extends DefaultTask {
    @Input
    abstract String getMetalavaBinaryPath();

    @InputFile
    abstract File getApiTxt();

    @Input
    abstract String getSourcePath();

    @InputFile
    abstract File getBaselineFile();

    @Input
    abstract boolean getUpdateBaseline();

    @Input
    abstract enum getMode();


    @OutputFile
    abstract File getOutputFile();

    abstract void setSourcePath(String value);

    abstract void setBaselineFile(File value);

    abstract void setUpdateBaseline(boolean value);

    abstract void setMetalavaBinaryPath(String value);

    abstract void setApiTxt(File value);

    abstract void setOutputFile(File value);

    abstract void setMode(Mode mode);

    private static enum Mode {
        GENERATE, COMPARE
    }

    @TaskAction
    void execute() {
        File outputFileDir = getOutputFile().getParentFile();
        if(!outputFileDir.exists()) {
            outputFileDir.mkdirs();
        }
        
        String cmd = mode == Mode.GENERATE ? "--check-compatibility:api:current %s" : "--api %s"
        String baseline = "--baseline %s" : "--update-baseline %s"
        String format = "--format=v2"
        String sourcePath = "--source-path %s"

        String cmdTemplate = ["%s", sourcePath, cmd, baselineArg, "--no-color"].join(" ")

        String cmdToRun = String.format(cmdTemplate, getMetalavaBinaryPath(), getSourcePath(), getApiTxt().getAbsolutePath(), getBaselineFile().getAbsolutePath());

        getProject().exec(spec-> {
            spec.setCommandLine(Arrays.asList(cmdToRun.split(" ")));
            try {
                spec.setStandardOutput(new FileOutputStream(getOutputFile()));
            } catch (FileNotFoundException e) {
                getLogger().error(e.toString());
            }
        });


        
    }

}
```

VinayGuthal replied:
in the ci command we would need to run the task to getApiInformation
and then run the task to generateApiInformation in this order
so we would need two tasks. However we can merge the two task classes into one though

VinayGuthal replied:
And there is no purpose of outputFile in the generateApiInformation Task

VinayGuthal replied:
The structure of the second task has changed now. It involves producing the api.txt file and then comparing the already existing api.txt with that. I feel now the structure of the two tasks are fairly different

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +63,51 @@ public void apply(Project project) {
                 }
               });
     }
+    if (System.getenv().containsKey("METALAVA_BINARY_PATH")) {
```
pls extract all this logic in a separate method with a descriptive name, additionally avoid nesting logic in `if`s like you're doing here, instead consider

```
if (!foo) { return; }
// stuff
if (!bar) { return; }
// other stuff.
```

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +63,51 @@ public void apply(Project project) {
+      }
+      if(doesSourcePathExist) {
+        project.getTasks().register("apiInformation", ApiInformationTask.class, task -> {
+          task.setProperty("apiTxt", project.file("api.txt"));
```
prefer using setters directly instead of using this "reflective magic".(here and below)

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -54,6 +63,51 @@ public void apply(Project project) {
+          task.setProperty("metalavaBinaryPath", metalavaBinaryPath);
+          task.setProperty("sourcePath", sourcePathArgument);
+          task.setProperty("outputFile", outputFile);
+          task.setProperty("baselineFile", project.file("baseline.txt"));
```
Instead of having empty baseline.txt files, can you pass it into metalava if it exists?

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java:
```diff
@@ -0,0 +1,94 @@
+        if(!outputFileDir.exists()) {
+            outputFileDir.mkdirs();
+        }
+        String cmdTemplate = getUpdateBaseline() ?
```
Since exec take a list, it looks like you don't need this "template" and can instead pass each individual item into `setCommandLine`. here and in the other task

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java:
```diff
@@ -0,0 +1,94 @@
+            try {
+                spec.setStandardOutput(new FileOutputStream(getOutputFile()));
+            } catch (FileNotFoundException e) {
+                getLogger().error(e.toString());
```
You should instead fail the build imo, by e.g. wrapping e in `GradleException`

vkryachko commented on firebase-database/baseline.txt:
```diff
@@ -0,0 +1,9 @@
+// Baseline format: 1.0
+RemovedInterface: com.google.firebase.database.Exclude:
```
I don't understand these lines, all of these are annotations so clearly they implement `Annotation`, also what does it compare it to? Removed when?

vkryachko commented on firebase-firestore/baseline.txt:
```diff
@@ -0,0 +1,21 @@
+    Attempted to remove @NonNull annotation from parameter elements in com.google.firebase.firestore.FieldValue.arrayUnion(Object... elements)
+
+
+RemovedInterface: com.google.firebase.firestore.DocumentId:
```
Same here

vkryachko commented on firebase-storage/baseline.txt:
```diff
@@ -0,0 +1,3 @@
+// Baseline format: 1.0
+RemovedInterface: com.google.firebase.storage.StorageException.ErrorCode:
+    Class com.google.firebase.storage.StorageException.ErrorCode no longer implements java.lang.annotation.Annotation
```
Here as well, I wonder if api.txt is produced in a format that does not understand annotations? or some input flags are missing? might be worth asking metalava folks

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/apiinfo/ApiInformationTask.java:
```diff
@@ -0,0 +1,94 @@
+            outputFileDir.mkdirs();
+        }
+        String cmdTemplate = getUpdateBaseline() ?
+                "%s --source-path %s --check-compatibility:api:current %s --format=v2 --update-baseline %s --no-color"
```
As I mentioned before you are still not populating the classpath

VinayGuthal commented:
The public api surface has changed for the subproject firebase-common:
error: Class com.google.firebase.FirebaseNetworkException superclass changed from com.google.firebase.FirebaseException to java.lang.Object [ChangedSuperclass]
error: Class com.google.firebase.FirebaseTooManyRequestsException superclass changed from com.google.firebase.FirebaseException to java.lang.Object [ChangedSuperclass]metalava wrote updated baseline to /usr/local/google/home/vguthal/firebase-android-sdk/firebase-common/baseline.txt


The public api surface has changed for the subproject firebase-database:
error: Added class com.google.firebase.database.DataSnapshot [AddedClass]metalava wrote updated baseline to /usr/local/google/home/vguthal/firebase-android-sdk/firebase-database/baseline.txt


Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-database:
warning: Added method com.google.firebase.database.ChildEventListener.onChildRemoved(com.google.firebase.database.DataSnapshot) [AddedAbstractMethod]

Please update the api.txt files for the subprojects being affected by this change with the files present in the artifacts directory. Also perform a major/minor bump accordingly.

VinayGuthal commented:
/test device-check-changed

vkryachko commented on buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.java:
```diff
@@ -71,6 +90,55 @@ public void apply(Project project) {
+    boolean doesSourcePathExist = false;
+    for(String sourcePath : sourcePathArgument.split(":")) {
+      if(new File(sourcePath).exists()) {
+        doesSourcePathExist = true;
```
I think you can replace this for loop with `mainSourceSet.getJava().getSrcDirs().stream().anyMatch(File::exists)`

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
error: Removed class com.google.firebase.firestore.FirebaseFirestore.InstanceRegistry [RemovedInterface]

The public api surface has changed for the subproject firebase-common:
error: Added field com.google.firebase.FirebaseApp.vinay [AddedField]
error: Added field com.google.firebase.FirebaseApp.vinay1 [AddedField]

Please update the api.txt files for the subprojects being affected by this change by running ./gradlew ${subproject}:generateApiTxtFile. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
error: Removed class com.google.firebase.firestore.FirebaseFirestore.InstanceRegistry [RemovedInterface]

The public api surface has changed for the subproject firebase-common:
error: Added field com.google.firebase.FirebaseApp.vinay [AddedField]
error: Added field com.google.firebase.FirebaseApp.vinay1 [AddedField]

Please update the api.txt files for the subprojects being affected by this change by running ./gradlew ${subproject}:generateApiTxtFile. Also perform a major/minor bump accordingly.

VinayGuthal commented:
The public api surface has changed for the subproject firebase-firestore:
error: Removed class com.google.firebase.firestore.FirebaseFirestore.InstanceRegistry [RemovedInterface]

The public api surface has changed for the subproject firebase-common:
error: Added field com.google.firebase.FirebaseApp.vinay [AddedField]
error: Added field com.google.firebase.FirebaseApp.vinay1 [AddedField]

Please update the api.txt files for the subprojects being affected by this change by running ./gradlew ${subproject}:generateApiTxtFile. Also perform a major/minor bump accordingly.
